go/tasks/plugins/array/k8s/monitor.go

```diff
@@ -24,73 +23,100 @@ import (
-	newArrayStatus := arraystatus.ArrayStatus{
+	newArrayStatus := &arraystatus.ArrayStatus{
 		Summary:  arraystatus.ArraySummary{},
 		Detailed: arrayCore.NewPhasesCompactArray(uint(currentState.GetExecutionArraySize())),
 	}
 
+	if int64(currentState.GetExecutionArraySize()) > config.MaxArrayJobSize {
+		ee := fmt.Errorf("array size > max allowed. Requested [%v]. Allowed [%v]", currentState.GetExecutionArraySize(), config.MaxArrayJobSize)
+		logger.Info(ctx, ee)
+		currentState = currentState.SetPhase(arrayCore.PhasePermanentFailure, 0).SetReason(ee.Error())
+		return currentState, logLinks, nil
+	}
+
+	// If we have arrived at this state for the first time then currentState has not been
+	// initialized with number of sub tasks.
+	if len(currentState.GetArrayStatus().Detailed.GetItems()) == 0 {
+		currentState.ArrayStatus = *newArrayStatus
+	}
+
 	for childIdx, existingPhaseIdx := range currentState.GetArrayStatus().Detailed.GetItems() {
 		existingPhase := core.Phases[existingPhaseIdx]
+		indexStr := strconv.Itoa(childIdx)
+		podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
 		if existingPhase.IsTerminal() {
 			// If we get here it means we have already "processed" this terminal phase since we will only persist
 			// the phase after all processing is done (e.g. check outputs/errors file, record events... etc.).
+
+			// Since we know we have already "processed" this terminal phase we can safely deallocate resource
+			err = deallocateResource(ctx, tCtx, config, childIdx)
+			if err != nil {
+				logger.Errorf(ctx, "Error releasing allocation token [%s] in LaunchAndCheckSubTasks [%s]", podName, err)
+				return currentState, logLinks, errors2.Wrapf(ErrCheckPodStatus, err, "Error releasing allocation token.")
+			}
 			newArrayStatus.Summary.Inc(existingPhase)
 			newArrayStatus.Detailed.SetItem(childIdx, bitarray.Item(existingPhase))
 
 			// TODO: collect log links before doing this
 			continue
 		}
 
-		phaseInfo, err := CheckPodStatus(ctx, kubeClient,
-			k8sTypes.NamespacedName{
-				Name:      formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), strconv.Itoa(childIdx)),
-				Namespace: tCtx.TaskExecutionMetadata().GetNamespace(),
-			})
-		if err != nil {
-			return currentState, logLinks, errors.Wrapf(ErrCheckPodStatus, err, "Failed to check pod status")
+		task := &Task{
+			LogLinks:         logLinks,
+			State:            newState,
+			NewArrayStatus:   newArrayStatus,
+			Config:           config,
+			ChildIdx:         childIdx,
+			MessageCollector: &msg,
 		}
 
-		if phaseInfo.Info() != nil {
-			logLinks = append(logLinks, phaseInfo.Info().Logs...)
+		// The first time we enter this state we will launch every subtask. On subsequent rounds, the pod
+		// has already been created so we return a Success value and continue with the Monitor step.
+		var status TaskStatus
+		status, err = task.Launch(ctx, tCtx, kubeClient)
```

Comment:
sorry if i'm missing something obvious here, but if the task is not terminal we always launch? what if it's already running?

Reply:
Yes that is correct. If it's already running we will get `alreadyExists` error and proceed with the rest of the flow in monitor.

Reply:
thanks for explaining